redash/models.py

```diff
@@ -96,6 +106,131 @@ def get(self, query_id):
 
 scheduled_queries_executions = ScheduledQueriesExecutions()
 
+
+class UTCDateTimeField(walrus.DateTimeField):
+    """
+    A walrus DateTimeField that makes the value timezone aware
+    using the pytz.utc timezone on return.
+    """
+    def python_value(self, value):
+        value = super(UTCDateTimeField, self).python_value(value)
+        if value.tzinfo is None:
+            value = value.replace(tzinfo=pytz.utc)
+        return value
+
+
+class UserDetail(walrus.Model):
+    """
+    A walrus data model to store some user data to Redis to be
+    synced to Postgres asynchronously.
+    """
+    __database__ = walrus_db
+    __namespace__ = 'redash.user.details'
+
+    user_id = walrus.IntegerField(index=True)
+    updated_at = UTCDateTimeField(index=True, default=utcnow)
+
+    @classmethod
+    def update(cls, user_id):
+        """
+        Update the user details hash using the given redis
+        pipeline, user id, optional redis id and optional user
+        details.
+
+        The fields uid, rid and updated (timestamp) are
+        enforced and can't be overwritten.
+        """
+        # try getting the user detail with the given user ID
+        # or create one if it doesn't exist yet (e.g. when key was purged)
+        try:
+            user_detail = cls.get(cls.user_id == user_id)
+            # update the timestamp with the current time
+            user_detail.updated_at = utcnow()
+            # save to Redis
+            user_detail.save()
+        except ValueError:
+            user_detail = cls.create(
+                user_id=user_id,
+                updated_at=utcnow(),
+            )
+        return user_detail
+
+    @classmethod
+    def sync(cls, chunksize=1000):
+        """
+        Syncs user details to Postgres (to the JSON field User.details).
+        """
+        to_sync = {}
+        try:
+            for user_detail in cls.all():
+                to_sync[user_detail.user_id] = user_detail
+
+            user_ids = list(to_sync.keys())
+            if not user_ids:
+                return
+            logger.info(
+                'syncing users: %s',
+                ', '.join([str(uid) for uid in user_ids])
+            )
+            # get all SQLA users that need to be updated
+            users = User.query.filter(User.id.in_(user_ids))
+            for i, user in enumerate(users):
+                update = to_sync[user.id]
+                updated_at = getattr(update, 'updated_at', None)
+                if updated_at is None:
```

Comment:
What this is protecting from?

Reply:
Ah, that's a left-over from the previous cases where we didn't delete user details objects and key revocation policy might have deleted the value of the individual walrus field. This is moot now with the fact we delete the whole set of values from Redis.

Reply:
Fixed.